File: grpc/cpp/sender_async_test.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <memory>
#include <future>
#include <grpcpp/grpcpp.h>
#include "messaging.grpc.pb.h"
#include "../../utils/cpp/stats_collector.hpp"
#include "../../utils/cpp/test_data_loader.hpp"
#include "../../utils/cpp/message_helpers.hpp"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using messaging::MessageEnvelope;
using messaging::MessagingService;
using json = nlohmann::json;
using message_helpers::get_current_time_ms;

struct TaskResult {
    bool success;
    std::string message_id;
    long long duration;
    std::string error;
};

TaskResult send_message_task(const json& item, int receiver_count) {
    TaskResult res;
    res.success = false;
    res.message_id = message_helpers::extract_message_id(item);
    res.duration = 0;
    
    try {
        int target = item.value("target", 0);
        int port = 50051 + target;
        
        auto channel = grpc::CreateChannel("localhost:" + std::to_string(port), grpc::InsecureChannelCredentials());
        auto stub = MessagingService::NewStub(channel);
        
        long long msg_start = get_current_time_ms();
        
        // Create protobuf message with DataMessage payload
        MessageEnvelope request = message_helpers::create_data_envelope(item);
        
        MessageEnvelope reply;
        ClientContext context;
        context.set_deadline(std::chrono::system_clock::now() + std::chrono::milliseconds(100));
        
        Status status = stub->SendMessage(&context, request, &reply);
        
        if (status.ok()) {
            if (message_helpers::is_valid_ack(reply, res.message_id)) {
                res.duration = get_current_time_ms() - msg_start;
                res.success = true;
            } else {
                res.error = "Invalid ACK";
            }
        } else {
            res.error = status.error_message();
        }
    } catch (const std::exception& e) {
        res.error = e.what();
    }
    
    return res;
}

int main() {
    auto test_data = test_data_loader::loadTestData();
    
    // Find max target
    int max_target = 0;
    for (const auto& item : test_data) {
        if (item.contains("target")) {
            max_target = std::max(max_target, item["target"].get<int>());
        }
    }
    
    MessageStats stats;
    stats.set_metadata({
        {"service", "gRPC"},
        {"language", "C++"},
        {"async", true}
    });
    long long start_time = get_current_time_ms();
    
    std::cout << " [x] Starting ASYNC transfer of " << test_data.size() << " messages..." << std::endl;
    
    std::vector<std::future<TaskResult>> futures;
    for (auto& item : test_data) {
        futures.push_back(std::async(std::launch::async, send_message_task, item, max_target + 1));
    }
    
    for (auto& fut : futures) {
        TaskResult res = fut.get();
        if (res.success) {
            stats.record_message(true, res.duration);
            std::cout << " [OK] Message " << res.message_id << " acknowledged" << std::endl;
        } else {
            stats.record_message(false);
            std::cout << " [FAILED] Message " << res.message_id << ": " << res.error << std::endl;
        }
    }
    
    long long end_time = get_current_time_ms();
    stats.set_duration(start_time, end_time);
    
    json report = stats.get_stats();
    
    std::cout << "\nTest Results (ASYNC):" << std::endl;
    std::cout << "total_sent: " << stats.sent_count << std::endl;
    std::cout << "total_received: " << stats.received_count << std::endl;
    std::cout << "duration_ms: " << stats.get_duration_ms() << std::endl;
    
    std::ofstream rf("logs/report.txt", std::ios::app);
    if (rf.good()) {
        rf << report.dump() << std::endl;
        rf.close();
    }
    
    return 0;
}

File: grpc/python/sender_async_test.py
#!/usr/bin/env python3
"""gRPC Python Sender - Async"""
import sys
import json
import asyncio
import grpc
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

import messaging_pb2
import messaging_pb2_grpc
from message_helpers import *
from test_data_loader import load_test_data
from stats_collector import MessageStats


async def send_message_task(stubs, item):
    """Send a single message asynchronously."""
    result = {'success': False, 'message_id': '', 'duration': 0, 'error': ''}
    
    try:
        message_id = extract_message_id(item)
        result['message_id'] = message_id
        target = item.get('target', 0)
        
        stub = stubs.get(target)
        if not stub:
             # Should create channel dynamically but for brevity assuming pre-created
             # In real async app, would manage pool
             result['error'] = 'No connection to target'
             return result
             
        msg_start = get_current_time_ms()
        
        envelope = create_data_envelope(item)
        
        response = await stub.SendMessage(envelope)
        
        if is_valid_ack(response, message_id):
            result['duration'] = get_current_time_ms() - msg_start
            result['success'] = True
        else:
            result['error'] = 'Invalid ACK'
            
    except grpc.RpcError as e:
        result['error'] = f"RPC Error: {e.code()}"
    except Exception as e:
        result['error'] = str(e)
    
    return result


async def run(base_port, num_receivers):
    test_data = load_test_data()
    
    stats = MessageStats()
    stats.set_metadata({
        'service': 'gRPC',
        'language': 'Python',
        'async': True
    })
    start_time = get_current_time_ms()
    
    print(f" [x] Starting ASYNC transfer of {len(test_data)} messages...")
    
    # Pre-create channels/stubs
    channels = []
    stubs = {}
    targets = set(item.get('target', 0) for item in test_data)
    
    for target in targets:
        port = base_port + target
        channel = grpc.aio.insecure_channel(f'localhost:{port}')
        stub = messaging_pb2_grpc.MessagingServiceStub(channel)
        channels.append(channel)
        stubs[target] = stub
    
    tasks = [send_message_task(stubs, item) for item in test_data]
    results = await asyncio.gather(*tasks)
    
    # Cleanup
    for channel in channels:
        await channel.close()
    
    for result in results:
        if result['success']:
            stats.record_message(True, result['duration'])
            print(f" [OK] Message {result['message_id']} acknowledged")
        else:
            stats.record_message(False)
            print(f" [FAILED] Message {result['message_id']}: {result['error']}")
    
    end_time = get_current_time_ms()
    stats.set_duration(start_time, end_time)
    
    report = stats.get_stats()
    
    print("\nTest Results (ASYNC):")
    print(f"total_sent: {stats.sent_count}")
    print(f"total_received: {stats.received_count}")
    print(f"duration_ms: {stats.get_duration_ms()}")
    
    with open('logs/report.txt', 'a') as f:
        f.write(json.dumps(report) + '\n')


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--base-port', type=int, default=50051)
    parser.add_argument('--num-receivers', type=int, default=1)
    args = parser.parse_args()
    
    asyncio.run(run(args.base_port, args.num_receivers))


if __name__ == "__main__":
    main()
